From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koji Ishii <kojii@chromium.org>
Date: Thu, 12 Sep 2024 05:51:00 +0000
Subject: [M126-LTS] Fix a range `CHECK` for when it overflows

This patch fixes a `CHECK` for a range of a string when
`offset + length` overflows the `unsigned`.

(cherry picked from commit 59c286e8419f07143ce859342f0fe9ddea36392d)

Bug: 355731798
Change-Id: If04222f10f2b73b6dcd6b412cf4d82fa5b71bbe2
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5776342
Commit-Queue: Kent Tamura <tkent@chromium.org>
Auto-Submit: Koji Ishii <kojii@chromium.org>
Commit-Queue: Koji Ishii <kojii@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1339526}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5804713
Reviewed-by: Koji Ishii <kojii@chromium.org>
Reviewed-by: Fahad Mansoor <fahadmansoor@google.com>
Auto-Submit: Roger Felipe Zanoni da Silva (xWF) <rzanoni@google.com>
Cr-Commit-Position: refs/branch-heads/6478@{#1958}
Cr-Branched-From: e6143acc03189c5e52959545b110d6d17ecd5286-refs/heads/main@{#1300313}

diff --git a/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.cc b/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.cc
index 5ece093127b61a334e271bc4a077e4f243307649..922b5d9c98e0750b8b20f5f7806723e8ae8990da 100644
--- a/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.cc
+++ b/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.cc
@@ -423,6 +423,12 @@ CanvasRotationInVertical CanvasRotationForRun(
 
 }  // namespace
 
+inline void HarfBuzzShaper::CheckTextLen(unsigned start,
+                                         unsigned length) const {
+  CHECK_LE(start, text_.length());
+  CHECK_LE(length, text_.length() - start);
+}
+
 void HarfBuzzShaper::CommitGlyphs(RangeContext* range_data,
                                   const SimpleFontData* current_font,
                                   UScriptCode current_run_script,
@@ -618,7 +624,7 @@ bool HarfBuzzShaper::CollectFallbackHintChars(
     if (it->action_ == kReshapeQueueNextFont)
       break;
 
-    CHECK_LE((it->start_index_ + it->num_characters_), text_.length());
+    CheckTextLen(it->start_index_, it->num_characters_);
     if (text_.Is8Bit()) {
       for (unsigned i = 0; i < it->num_characters_; i++) {
         hint.push_back(text_[it->start_index_ + i]);
diff --git a/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.h b/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.h
index cfa6b8c69a863d59c869fa138fa30491e54c480b..12ee16372ed23eb59df122b481dfaebc6b4fa779 100644
--- a/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.h
+++ b/third_party/blink/renderer/platform/fonts/shaping/harfbuzz_shaper.h
@@ -156,6 +156,8 @@ class PLATFORM_EXPORT HarfBuzzShaper final {
                     const BufferSlice&,
                     ShapeResult*) const;
 
+  void CheckTextLen(unsigned start, unsigned length) const;
+
   const String text_;
   EmojiMetricsCallback emoji_metrics_reporter_for_testing_;
 };
